File: EPI/8.4/swift/8.4.swift
public class Node {
    public var next: Node?

    // cycleStart returns the node at the start of the cycle, or nil if there is no cycle.
    public func cycleStart() -> Node? {
        if let node = self.cycleAny() {
            let count = node.cycleCount()
            return self.cycleStart(count)
        }
        return nil
    }

    // cycleAny returns any node of the cycle, or nil if there is no cycle.
    func cycleAny() -> Node? {
        var slow = self
        var fast = self.next

        while let f = fast, fnn = f.next?.next {
            if slow === f {
                return slow
            }
            slow = slow.next!
            fast = fnn
        }

        return nil
    }

    // cycleCount returns the number of nodes in the cycle containing the current node.
    func cycleCount() -> Int {
        var next = self.next!
        var count = 1

        while next !== self {
            count += 1
            next = next.next!
        }

        return count
    }

    // cycleStart returns the node at the start of the cycle containing cycleCount nodes.
    func cycleStart(cycleCount: Int) -> Node {
        // Use two cursors separated by cycleCount nodes. They will point to the same node after the
        // second cursor completes a full cycle.
        var firstCursor = self
        var secondCursor = self

        for _ in 0..<cycleCount {
            secondCursor = secondCursor.next!
        }

        while firstCursor !== secondCursor {
            firstCursor = firstCursor.next!
            secondCursor = secondCursor.next!
        }

        return firstCursor
    }
}

File: EPI/8.4/typescript/8.4.ts
export interface Node {
    next?: Node;
}

export function getFirstNodeInCycle(list: Node): Node | undefined {
    const node = getNodeInCycle(list);
    return node && getFirstNode(list, getCycleLength(node));
}

function getNodeInCycle(list: Node): Node | undefined {
    let slow = list;
    let fast = list;

    while (fast.next && fast.next.next) {
        slow = slow.next!;
        fast = fast.next.next;
        if (slow === fast) {
            return slow;
        }
    }

    return undefined;
}

function getCycleLength(node: Node): number {
    let length = 1;
    let n = node.next!;

    while (n !== node) {
        length += 1;
        n = n.next!;
    }

    return length;
}

function getFirstNode(list: Node, cycleLength: number): Node {
    let first = list;
    let second = list;

    for (let i = 0; i < cycleLength; i += 1) {
        second = second.next!;
    }

    while (first !== second) {
        first = first.next!;
        second = second.next!;
    }

    return first;
}
